sim-lib/src/lib.rs

```diff
@@ -988,64 +1045,79 @@ async fn produce_simulation_results(
             biased;
-            _ = shutdown.clone() => break,
+            _ = listener.clone() => {
+                log::debug!("Produce simulation results received shutdown signal. Simulation results producer exiting.");
+                return Ok(())
+            },
             output = output_receiver.recv() => {
-                match output{
+                match output {
                     Some(simulation_output) => {
                         match simulation_output{
                             SimulationOutput::SendPaymentSuccess(payment) => {
-                                let source_node = nodes.get(&payment.source).unwrap().clone();
-                                set.spawn(track_payment_result(
-                                    source_node, results.clone(), payment, shutdown.clone(),
-                                ));
+                                if let Some(source_node) = nodes.get(&payment.source) {
+                                    set.spawn(track_payment_result(
+                                        source_node.clone(), results.clone(), payment, listener.clone()
+                                    ));
+                                } else {
+                                    log::error!("Can't find source node with public key: {}.", payment.source);
+                                    return Err(SimulationError::MissingNodeError(format!("Source node with public key: {} unavailable.", payment.source)));
+                                }
                             },
                             SimulationOutput::SendPaymentFailure(payment, result) => {
-                                if results.clone().send((payment, result)).await.is_err() {
-                                    log::debug!("Could not send payment result.");
+                                if let Err(e) = results.send((payment, result.clone())).await {
+                                    log::error!("Could not send payment result.");
+                                    return Err(SimulationError::MpscChannelError{
+                                        msg: format!("Failed to send payment result: {result} for payment {:?} dispatched at {:?}.", payment.hash, payment.dispatch_time),
+                                        error_str: e.to_string()
+                                    });
                                 }
                             }
                         };
-
                     },
-                    None => {
-                        return
+                    None => return Ok(())
+                }
+            },
+            track_payment = set.join_next() => {
+                if let Some(res) = track_payment {
+                    match res {
+                        Ok(track_payment_res) => {
+                            if let Err(e) = track_payment_res {
+                                log::error!("Track payment result task exited");
+                                return Err(e);
```

Comment:
We log all of the errors returned by this function at the call site so we don't need to log each one as we return them. Applies in a few places.

Reply:
Done the cleanup at all locations.